File: src/model/Question.java
package model;

import java.util.List;
import java.util.Scanner;



public class Question {
	
	private static String INCORRECT_TEXT = "INCORRECT_ANSWER";
	
	private final  String myQuestionText;
	private final String myCorrectAnswer;
	private List<String> myChoices;
	private boolean isAskedAlready;
	public Question(String theQuestionText, String theCorrectAnswer, List<String> theChoices) {
		this.myQuestionText = setMyQuestionText(theQuestionText);
		this.myCorrectAnswer = setMyCorrectAnswer(theCorrectAnswer);
		this.myChoices = setMyChoices(theChoices);
		this.isAskedAlready = false;
	}
	
	private void invalidateChoice(int theInvalidChoiceIdx) {
		this.myChoices.set(theInvalidChoiceIdx, INCORRECT_TEXT);
	}
	
	
	// On the UML this was a string, however it makes sense to return a boolean for usage in other classes
	// ex: doors close upon wrong answers, which we can use a true false value for movement vs closing a path. 
	public boolean askQuestion() {
		this.printQuestion();
		this.printChoices();
		System.out.println("Enter an int to select a choice: ");
		
		Scanner scanny = new Scanner(System.in);
		int theChosenAnswerIdx = scanny.nextInt() -1 ;
		String theChosenAnswer = myChoices.get(theChosenAnswerIdx);
		boolean isCorrect = theChosenAnswer.equals(this.myCorrectAnswer);
		if(!isCorrect) {
			this.invalidateChoice(theChosenAnswerIdx);
		}
		
		this.printResult(isCorrect);
		return isCorrect;
	}
	
	public String getMyQuestionText() {
        return myQuestionText;
    }

    public String getMyCorrectAnswer() {
        return myCorrectAnswer;
    }

    public List<String> getMyChoices() {
        return myChoices;
    }
    
	private String setMyQuestionText(String theQuestionText) {
		if(!theQuestionText.equals(null) || theQuestionText.length() == 0) {
			return theQuestionText;
		}
		throw new IllegalArgumentException("Invalid Value: Question text was blank/null.");
	}
	
	private String setMyCorrectAnswer(String theQuestionAnswer) {
		if(!theQuestionAnswer.equals(null) || theQuestionAnswer.length() == 0) {
			return theQuestionAnswer;
		}
		throw new IllegalArgumentException("Invalid Value: Answer text was blank/null.");
	}
	
	private List<String> setMyChoices(List<String> theChoices) {
		if(!theChoices.equals(null) || theChoices.size() == 0) {
			return theChoices;
		}
		throw new IllegalArgumentException("Invalid Value: Choices list was blank/null.");
	}
	
	public void setIsAskedAlready(boolean theStatus) {
		this.isAskedAlready = theStatus;
	}
	
	private void printQuestion() {
		System.out.println(this.getMyQuestionText());
	}
	
	private void printChoices() {
		List<String> theChoices = this.getMyChoices();
		int count = 1;
		for(String choice: theChoices) {
			System.out.println(count + ": " + choice);
			count++;
		}
	}
	
	private void printResult(boolean result) {
		if(result) {
	        System.out.println("Correct!");
	    } else {
	        System.out.println("Incorrect!");
	    }
	}
	
}

File: vite-project/src/models/Question.ts
import Item from "./Item";

export default class Question{
    public static INCORRECT_TEXT = "INCORRECT_ANSWER";

    private myQuestionText!: string;
    private myCorrectAnswer!: string;
    private myChoices!: string[];

    constructor(theQuestionText: string, theCorrectAnswer: string, theChoices: string[]) {
        this.myQuestionText = this.setMyQuestionText(theQuestionText);
        this.myCorrectAnswer = this.setMyCorrectAnswer(theCorrectAnswer);
        this.myChoices = this.setMyChoices(theChoices);
    }

    public applyItem(theItem: Item): void {
        let itemType = theItem.getItemType();
        if(itemType === "50-50"){
            this.applyFiftyFifty();
        } else if(itemType === "Phone-a-Friend"){
            this.applyPhoneAFriend();
        }
    }
    private wrongAnswer(): string {
        let randomIndex = Math.floor(Math.random() * this.myChoices.length);
        let wrongAnswers = this.myChoices.filter(choice => choice !== this.myCorrectAnswer);

        return wrongAnswers[randomIndex];
    }
    private applyPhoneAFriend(): void {
        let chance = Math.random();

        
       
        if(chance < 0.1) {
            alert( this.wrongAnswer());
        } else {
            alert(this.myCorrectAnswer);
        }
    }

    private applyFiftyFifty(): void {
        if(this.myChoices.includes(Question.INCORRECT_TEXT)){
            console.log("Fifty fifty can only be applied on the first turn.")
            return
        }
        let elim = 0

        while(elim < 2) {
            let wrongAnswer = this.wrongAnswer();
            if(wrongAnswer !== Question.INCORRECT_TEXT){
                this.myChoices.splice(this.myChoices.indexOf(wrongAnswer), 1);
                elim++;
            }
        }
        console.log("Fifty fifty eliminated 2 choices")
    }

    private invalidateChoice(theInvalidChoice:number): void {
        this.myChoices[theInvalidChoice] = Question.INCORRECT_TEXT;
    }

    public askQuestion():boolean{
        this.printQuestion();
        this.printChoices();

        let answer = prompt("Enter the number of your answer: ");
        if(answer === null){
            console.log("Not a valid answer. Please enter a number");
            return false;
        }
        let answerNumber = parseInt(answer)-1;
        let theChosenAnswer = this.myChoices[answerNumber];
        let isCorrect = theChosenAnswer === this.myCorrectAnswer;
        if(!isCorrect){
            this.invalidateChoice(answerNumber);
        }

        this.printResult(isCorrect);
        return isCorrect;
    }
    
    public getMyQuestionText(): string {
        return this.myQuestionText;
    }

    public getMyCorrectAnswer(): string {
        return this.myCorrectAnswer;
    } 

    public getMyChoices(): string[] {
        return this.myChoices;
    }   

    public setMyQuestionText(theQuestionText: string): string {
        if(theQuestionText.length === 0) {
            throw new Error(Question.INCORRECT_TEXT);
        }
       return theQuestionText;
    }

    public setMyCorrectAnswer(theCorrectAnswer: string): string {
        if(theCorrectAnswer.length === 0) {
            throw new Error(Question.INCORRECT_TEXT);
        }
        return theCorrectAnswer;
    }

    public setMyChoices(theChoices: string[]): string[] {
        if(theChoices.length === 0) {
            throw new Error(Question.INCORRECT_TEXT);
        }
        return theChoices;
    }

    private printQuestion(): void {
        console.log(this.myQuestionText);
    }

    private printChoices(): void {
        let choiceNumber = 1;
        this.myChoices.forEach(choice => {
            console.log(choiceNumber + ". " + choice);
            choiceNumber++;
        });
    }

    private printResult(result: boolean): void {
        if(result) {
            console.log("Correct!");
        } else {
            console.log("Incorrect!");
        }
    }
    
    public toString(): string {
        return this.myQuestionText;
    }    
}